### DISCUSSION TITLE
NativeMenuFlyout support (native context menu)

### DISCUSSION THREAD

**Discussion by maxkatz6** [AvaloniaUI/Avalonia#8581] [Ideas]
Currently we have somewhat strict Flyout API, that is not tied to the VisualTree content (like ContextMenu or Popup) and can be potentially extended to support native menus.
If native menu API is not available, it should fallback to a popup with no issue, bridging NativeMenuItem to a MenuItem.
At the same time, we can freely extend popup flyout API with popup and controls related members.
Before hands platform specific APIs should be reviewed.

New API, reusing NativeMenu container with its lookless items tree:
```diff
+public class NativeMenuFlyout : FlyoutBase
+{
+    [Content]
+    public NativeMenu Menu { get; }
+    public ControlTheme FallbackMenuFlyoutPresenterTheme { get; set; }
+}
```

Typical usage:
```xaml
<TextBox Name="MyTextBox">
   <TextBox.ContextFlyout>
       <NativeMenuFlyout>
           <NativeMenu>
                <NativeMenuItem Header="Copy" Command="{Binding #MyTextBox.Copy}" />
                <NativeMenuItem Header="Cut" Command="{Binding #MyTextBox.Cut}" />
                <NativeMenuItem Header="Paste" Command="{Binding #MyTextBox.Paste}" />
           </NativeMenu>
       </NativeMenuFlyout>
   <TextBox.ContextFlyout>
</TextBox>
```

TODO:
Add macOS/apple specific attached properties for integration with `UITextInteraction` and/or global macOS hotkeys.
Something like `<NativeMenuItem Header="Writing Tools" MacOSProperties.WritingToolsContainer="True" />` ?

Relevant changes done in 11.0:
```diff
public class FlyoutBase
{
    bool IsOpen { get; }
    FlyoutPlacementMode Placement { get; set; } // Most likely not all options can be supported with the native API
    event EventHandler? Closed;
    event EventHandler? Opened;
-   FlyoutShowMode ShowMode { get; set; }
-   EventHandler<CancelEventArgs>? Closing;
-   EventHandler? Opening;

    void ShowAt(Control placementTarget);
}

+public class PopupFlyoutBase
+{
+    FlyoutShowMode ShowMode { get; set; } // Moved from FlyoutBase
+    EventHandler<CancelEventArgs>? Closing; // Moved from FlyoutBase
+    EventHandler? Opening; // Moved from FlyoutBase
+
+    double HorizontalOffset { get; set; } // New property, although it might be possible to use with native menu with no issue
+    double VerticalOffset { get; set; } // New property
+}

-public class MenuFlyout : FlyoutBase
+public class MenuFlyout : PopupFlyoutBase
{
   public IEnumerable<MenuItem> Items { get; } 
}
```

## Platform support

<details>
<summary>Windows UWP.MenuFlyout (XAML Island)</summary>

| ShowAt(pos) | AnchorRectangle | Anchor | Gravity | ConstraintAdjustment | ShowMode | Offset |
|---|---|---|---|---|---|---|
| ✔ | ⚡ | ⚡ | ⚡ | ⚡ | ✔ | ✔ |

| Opened | Opening (cancellable) | Closed | Closing (cancellable) |
|---|---|---|---|
| ✔ | ❌ | ✔ | ✔ |

| Nested Items | Item.Icon | Item.Gesture | Item.Toggle type | Item.Radio type |
|---|---|---|---|---|
| ✔ | ✔ | ✔ | ✔ | ✔ |

Notes:
- Requires extra changes to the app.manifest 
- Pretty complicated COM interop for all of the requirements
- Advanced positioning and anchoring possibly can be emulated by creating fake overlay uwp visual

![image](https://github.com/user-attachments/assets/dadf0bd8-9b80-46f7-8533-45ebff2635f1)

</details>

<details>
<summary>Windows CreatePopupMenu</summary>

TODO

We might prefer CreatePopupMenu over UWP.MenuFlyout for simplicity and TrayIcon integration

![image](https://github.com/user-attachments/assets/1ffde4aa-64d9-4ea3-ade2-066ca6e45fad)

</details>

<details>
<summary>Android PopupMenu</summary>

| ShowAt(pos) | AnchorRectangle | Anchor | Gravity | ConstraintAdjustment | ShowMode | Offset |
|---|---|---|---|---|---|---|
| ❌ | ⚡ | ⚡ | ⚡ | ⚡ | ❌ | ❌ |

| Opened | Opening (cancellable) | Closed | Closing (cancellable) |
|---|---|---|---|
| ✔ | ✔ | ✔ | ❌ |

| Nested Items | Item.Icon | Item.Gesture | Item.Toggle type | Item.Radio type |
|---|---|---|---|---|
| ⚡ | ⚡ | ⚡ | ✔ | ❌ |

Notes:
- Must be a native View rect for AnchorRectangle.
- Limited with single Gravity enum for Anchor, Gravity, and ConstraintAdjustment.
- Only one sub level for nested items.
- Only top-level icons supported.
- Can use PopupWindow for the ShowAt(pos)

![image](https://github.com/user-attachments/assets/edf49575-5057-4116-9ee9-3516d13c7f4c)

</details>

<details>
<summary>iOS UIAlertController</summary>

| ShowAt(pos) | AnchorRectangle | Anchor | Gravity | ConstraintAdjustment | ShowMode | Offset |
|---|---|---|---|---|---|---|
| ❌ | ⚡ | ❌ | ⚡ | ❌ | ❌ | ❌ |

| Opened | Opening (cancellable) | Closed | Closing (cancellable) |
|---|---|---|---|
| ✔ | ✔ | ✔ | ❌ |

| Nested Items | Item.Icon | Item.Gesture | Item.Toggle type | Item.Radio type |
|---|---|---|---|---|
| ❌ | ❌ | ❌ | ❌ | ❌ |

Notes:
- It's more of a dialog rather than a menu
- AnchorRectangle only supported on iPad. Technically can show nested sheets by closing parent one.

![image](https://github.com/user-attachments/assets/250660f6-8b6e-41f1-84b2-036fbfca3076)

</details>

<details>
<summary>iOS UIMenu</summary>

TODO: might be a better option than UIAlertController, and also has up to date styling, generally looks like a proper menu

Note:
- Requires fake button to get positioning
- Can ONLY use it programmatically on 17.0+ with a fake button, see [performPrimaryAction](https://developer.apple.com/documentation/uikit/uicontrol/performprimaryaction()?language=objc)
- OR by [faking touchesBegan](https://stackoverflow.com/a/76744276)
- OR by [using undocumented API](https://github.com/xybp888/iOS-Header/blob/master/13.0/PrivateFrameworks/UIKitCore.framework/UIContextMenuInteraction.h#L85)

![image](https://github.com/user-attachments/assets/d23d3466-0260-4bc9-85ed-5a7cc4399e1e)

</details>

<details>
<summary>macOS NSMenu + popUp</summary>

| ShowAt(pos) | AnchorRectangle | Anchor | Gravity | ConstraintAdjustment | ShowMode | Offset |
|---|---|---|---|---|---|---|
| ⚡ | ❌ | ❌ | ❌ | ❌ | ❌ | ❌ |

| Opened | Opening (cancellable) | Closed | Closing (cancellable) |
|---|---|---|---|
| ✔ | ✔ | ✔ | ❌ |

| Nested Items | Item.Icon | Item.Gesture | Item.Toggle type | Item.Radio type |
|---|---|---|---|---|
| ✔ | ✔ | ✔ | ✔ | ✔ |

![image](https://github.com/user-attachments/assets/88bb640a-4c12-436f-a688-41a186f13079)

</details>

<details>
<summary>Linux</summary>

None, except using GTK/QT which is unlikely.

</details>

<details>
<summary>Browser</summary>

Single proposed context menu API existed only in Mozilla and has been removed since
</details>

**Comment by timunie** [comment]
First of all I think this idea is really good and not too hard to implement, so I vote for yes 👍. 

--------------

**Additional** 
If we are already talking about modifying this API I would like to add another idea. Atm if Placement is Down and the space is not enough, the Popup overlays the PlacementTarget. This can make your selected item in a ComboBox invisible. My idea here would be to add a bool if the placement can switch in that case, so show above instead below like WPF does it. If it's not related, please tell me. I can open another issue then.

**Reply by kekekeks** [reply]
Right now flyout placement modes seems to be some hack to make the api somewhat compatible with UWP.
1) We don't pass the proper placement rect to the popup and instead do this - https://github.com/AvaloniaUI/Avalonia/blob/49aad04861f19c02ddc6a0780750ad29cc474c3b/src/Avalonia.Controls/Flyouts/FlyoutBase.cs#L437-L515
2) We don't expose our _way richer_ popup positioning configuration API: https://github.com/AvaloniaUI/Avalonia/blob/49aad04861f19c02ddc6a0780750ad29cc474c3b/src/Avalonia.Controls/Primitives/PopupPositioning/IPopupPositioner.cs#L77-L253
https://github.com/AvaloniaUI/Avalonia/blob/49aad04861f19c02ddc6a0780750ad29cc474c3b/src/Avalonia.Controls/Primitives/PopupPositioning/IPopupPositioner.cs#L308-L422

**Reply by timunie** [reply]
Does it mean we *should* improve it or does it mean it will not work?

**Reply by kekekeks** [reply]
`FlyoutPlacementMode` values should be merged into `PlacementMode`. Additional properties from `Popup` should be exposed on the `FlyoutBase`

**Reply by kekekeks** [reply]
Regarding your combobox case: it can be handled via Gravity.BottomRight + Anchor.BottomLeft + ConstraintAdjustment.FlipX/FlipY

**Reply by timunie** [reply]
Thank you for the suggestion ❤️

**Comment by kekekeks** [comment]
```
public class NativeMenuFlyout : FlyoutBase
{
    public IEnumerable<NativeMenuItem> Items { get; }
}
```

I think it should be `NativeMenu` as a container instead since we'll be adding `NativeMenu.Items` property that supports data templates

**Comment by maxkatz6** [comment]
Added some information about native popups compatibility and features

**Reply by timunie** [reply]
We may ask the one who is doing the PR #8352 if he knows more about popups in Wayland. 

//cc @affederaffe
